### PR TITLE
SI-8708 Fix pickling of LOCAL_CHILD child of sealed classes

### PR DESCRIPTION
**Pull request by lrytz**
When a sealed class or trait has local children, they are not pickled
in as part of the children of the symbol (introduced in 12a2b3b to fix
Aladdin bug 1055). Instead the compiler adds a single child class
named LOCAL_CHILD. The parents of its ClassInfoType were wrong: the
first parent should be a class. For sealed traits, we were using the
trait itself.

Also, the LOCAL_CHILD dummy class was entered as a member of its
enclosing class, which is wrong: it represents a local (non-member)
class, and it's a synthetic dummy anyway.

### CODE REVIEW

**Comment by lrytz** [comment]
review by @adriaanm

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala [hunk 1]
```diff
@@ -290,6 +290,25 @@ abstract class UnPickler {
...
       def finishSym(sym: Symbol): Symbol = {
+        /**
+         * member symbols (symbols owned by a class) are added to the class's scope, with a number
+         * of exceptions:
+         *
+         * (.) ...
+         * (1) `local child` represents local child classes, see comment in Pickler.putSymbol.
+         *     Since it is not a member, it should not be entered in the owner's scope.
+         */
+        def shouldEnterInOwnerScope = {
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:301]: This looks a lot like the negation of `isLocalToPickle` -- can we consolidate?

```
    private def isLocalToPickle(sym: Symbol): Boolean = (sym != NoSymbol) && !sym.isPackageClass && (
         isRootSym(sym)
      || sym.isRefinementClass
      || sym.isAbstractType && sym.hasFlag(EXISTENTIAL) // existential param
      || sym.isParameter
      || isLocalToPickle(sym.owner)
    )
```

**Comment by adriaanm** [comment]
That's some thorough testing!

File: test/files/run/t8708_b/Test_2.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import scala.tools.partest._
+import java.io.{Console => _, _}
+
+object Test extends DirectTest {
+
+  override def extraSettings: String = "-usejavacp -cp " + testOutput.path
+
+  override def code = "class Foo { val f = (param: Int) => param }"
```
Comment (lrytz) [inline test/files/run/t8708_b/Test_2.scala]: This is not necessary, should be removed.

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala (hunk 1 above)
Comment (lrytz) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:301]: I think the two answer different questions. `isLocalToPickle` decides whether a symbol is declared within the class that's currently being pickled. For example, `class C[T] { val x: Int }`, the symbols for `x` and `T` are both local to pickle.

`shouldEnterInOwnerScope` on the other hand answers the question if the unpickled symbol should be entered in scope. This is true for `x`, but not for `T`.

Comment (adriaanm) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:301]: But if you were to draw a venn diagram (I like to follow this [advice](http://www.nbr.co.nz/article/10-tricks-appear-smart-during-meetings-ck-158786)), isn't there an intersection that's intentionally non-empty?

Concretely, should one of these imply the other? Should they be kept in synch then?

```
      || sym.isRefinementClass
      || sym.isAbstractType && sym.hasFlag(EXISTENTIAL) // existential param
      || sym.isParameter
```

Rewriting the code here a bit:

```
!(
      || sym.isRefinementClass
      || sym.isExistentiallyBound
      || sym.isTypeParameter // implies sym.isParameter
      || sym.rawname == tpnme.LOCAL_CHILD
)
```

**Comment by adriaanm** [comment]
Looks like issues with the network on our side. The repo did not log any errors.

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala (hunk 1 above)
Comment (lrytz) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:301]: I see that the two have some tests in common, but logically they do different things. For example, we don't need a special case for `LOCAL_CHILD` in `isLocalToPickle`. I don't think it's more clear to share.

```
// not sure what this should be called, what it expresses
def common(sym) = ...
def isLocalToPickle(sym) = someTests && common(sym)
def shouldEnterInOwnerScope == someOtherTests && common(sym)
```

Comment (adriaanm) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:301]: Ok, fair enough. Thanks for explaining.

**Comment by lrytz** [comment]
PLS REBUILD ALL

**Comment by lrytz** [comment]
LGT @adriaanm ?

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**